### PR TITLE
Account touch trait

### PR DESCRIPTION
**Pull request by muharem**
This PR introduces `AccountTouch` trait and its implementation by `pallet-assets` which allows the creation of asset account for some account id with a designated depositor. 

requires https://github.com/paritytech/substrate/pull/13843

### CODE REVIEW

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2765453

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2765454

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2771090

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2771091

File: frame/assets/src/functions.rs [hunk 1]
```diff
@@ -308,17 +308,20 @@ impl<T: Config<I>, I: 'static> Pallet<T, I> {
 	}
 
 	/// Creates an account for `who` to hold asset `id` with a zero balance and takes a deposit.
+	/// If invoked `permissionless = true` than a depositor can be anyone, if permissioned
+	/// a depositor must be the asset's `Admin` or `Freezer`.
```
Comment (kianenigma) [inline frame/assets/src/functions.rs:312]:
```suggestion
	/// 
	/// If invoked `permissionless = true` than a depositor can be anyone, if `permissioned = false`
	/// a depositor must be the asset's `Admin` or `Freezer`.
```

is this what you meant?

File: frame/assets/src/lib.rs [hunk 2]
```diff
@@ -1601,6 +1600,21 @@ pub mod pallet {
 			Self::do_refund_other(id, &who, &origin)
 		}
 	}
+
+	/// Implements `AccountTouch` trait.
```
Comment (kianenigma) [inline frame/assets/src/lib.rs:1604]: Please use proper links, in light of the 4th principle named [here](https://forum.parity.io/t/new-polkadot-substrate-documentation-plan/1810) (still working on writing a guideline for how to document FRAME code, but one is indeed using links where possible).

File: frame/support/src/traits/account.rs [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+// limitations under the License.
+
+//! Traits for dealing with accounts.
+
+use crate::dispatch::DispatchResult;
+
+/// Trait for creating an asset account with a deposit taken from a specified by client depositor.
+pub trait Touch<AssetId, AccountId> {
+	/// The type for currency units of the deposit.
+	type Balance;
```
Comment (kianenigma) [inline frame/support/src/traits/account.rs:25]:
```suggestion
	type Balance;

```

File: frame/support/src/traits/account.rs [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+	/// The deposit amount of a native currency required for creating an asset account.
```
Comment (kianenigma) [inline frame/support/src/traits/account.rs:26]: Insufficient documentation IMO. What does `None` represent?

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2772879

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2772880

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2773229

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2773228

File: frame/support/src/traits/misc.rs [hunk 5]
```diff
@@ -1154,6 +1154,18 @@ impl<Hash> PreimageRecipient<Hash> for () {
 	fn unnote_preimage(_: &Hash) {}
 }
 
+/// Trait for creating an asset account with a deposit taken from a specified by client depositor.
+pub trait AccountTouch<AssetId, AccountId> {
+	/// The type for currency units of the deposit.
+	type Balance;
+
+	/// The deposit amount of a native currency required for creating an asset account.
+	fn deposit() -> Self::Balance;
```
Comment (gavofyork) [inline frame/support/src/traits/misc.rs:1163]: `deposit` is a verb as well as a noun; as a function you'll tend to read the verb form of a word over the noun-form.

Comment (gavofyork) [inline frame/support/src/traits/misc.rs:1163]:
```suggestion
	fn deposit_required() -> Self::Balance;
```

File: frame/assets/src/lib.rs [hunk 6]
```diff
@@ -1601,6 +1600,21 @@ pub mod pallet {
...
 		}
 	}
+
+	/// Implements [AccountTouch] trait.
+	/// Note that a depositor can be any account, without any specific privilege.
+	/// This implementation is supposed to be used only for creation of system accounts.
+	impl<T: Config<I>, I: 'static> AccountTouch<T::AssetId, T::AccountId> for Pallet<T, I> {
+		type Balance = DepositBalanceOf<T, I>;
+
+		fn deposit() -> Self::Balance {
```
Comment (gavofyork) [inline frame/assets/src/lib.rs:1610]:
```suggestion
		fn deposit_required() -> Self::Balance {
```

**Review by gavofyork — COMMENTED** [review]
Quibbles over docs and naming, but seems harmless otherwise

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2777683

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2777684

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2777940

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-benches
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2777941

File: frame/assets/src/functions.rs [hunk 7]
```diff
@@ -308,17 +308,21 @@ impl<T: Config<I>, I: 'static> Pallet<T, I> {
 	}
 
 	/// Creates an account for `who` to hold asset `id` with a zero balance and takes a deposit.
+	///
+	/// If invoked `permissionless = true` than a depositor can be anyone, if `permissioned = false`
+	/// a depositor must be the asset's `Admin` or `Freezer`.
```
Comment (joepetrowski) [inline frame/assets/src/functions.rs:313]:
```suggestion
	/// If invoked `permissionless = true`, then a depositor can be anyone. If `permissioned = false`,
	/// then the depositor must be the asset's `Admin` or `Freezer`.
```

**Review by joepetrowski — APPROVED** [review]

File: frame/assets/src/functions.rs [hunk 8]
```diff
@@ -308,17 +308,21 @@ impl<T: Config<I>, I: 'static> Pallet<T, I> {
 	}
 
 	/// Creates an account for `who` to hold asset `id` with a zero balance and takes a deposit.
+	///
+	/// If invoked `permissionless = true`, then a depositor can be anyone. If `permissioned = false`,
```
Comment (jsidorenko) [inline frame/assets/src/functions.rs:312]:
```suggestion
/// If invoked `permissionless = true`, then a depositor can be anyone. If `permissionless = false`,
```

**Pull request merged**